ksmt-core/src/main/kotlin/org/ksmt/cache/weak/ConcurrentWeakHashMapCache.kt

```diff
@@ -0,0 +1,321 @@
+            modificationsSinceLastCleanup.set(0)
+            scheduleCleanup()
+        } else {
+            runCleanupIfRequired()
+        }
+    }
+
+    private fun runCleanupIfRequired() {
+        if (cleanupStatus.get() == REQUIRED) {
+            runCleanup()
+        }
+    }
+
+    private fun scheduleCleanup() {
+        while (true) {
+            when (cleanupStatus.get()) {
+                // No ongoing cleanup -> schedule and run
+                IDLE -> {
+                    cleanupStatus.compareAndSet(IDLE, REQUIRED)
+                    runCleanup()
+                    return
+                }
+                // No ongoing cleanup and cleanup is already scheduled -> run
+                REQUIRED -> {
+                    runCleanup()
+                    return
+                }
+                // Cleanup is running. Try to reschedule cleanup after completion.
+                PROCESSING_TO_IDLE -> {
+                    if (cleanupStatus.compareAndSet(PROCESSING_TO_IDLE, PROCESSING_TO_REQUIRED)) {
+                        return
+                    }
+                    // Cleanup status changed. Retry
+                }
+                // Cleanup is running and will be rescheduled right after completion.
+                PROCESSING_TO_REQUIRED -> {
+                    return
+                }
+            }
+        }
+    }
+
+    private inline fun ifNotProcessing(body: () -> Unit) {
+        val status = cleanupStatus.get()
+        if (status == PROCESSING_TO_IDLE || status == PROCESSING_TO_REQUIRED) {
+            return
+        }
+        body()
+    }
+
+    private fun runCleanup() = ifNotProcessing {
+        if (cleanupLock.tryLock()) {
+            try {
+                ifNotProcessing {
+                    cleanup()
+                }
+            } finally {
+                cleanupLock.unlock()
+            }
+        }
+    }
+
+    private fun cleanup() {
+        cleanupStatus.set(PROCESSING_TO_IDLE)
+        try {
+            drainKeyReferences()
+        } finally {
+            if (!cleanupStatus.compareAndSet(PROCESSING_TO_IDLE, IDLE)) {
+                cleanupStatus.set(REQUIRED)
+            }
+        }
+    }
+
+    private fun drainKeyReferences() {
+        while (true) {
+            val keyRef = keyReferenceQueue.poll() ?: break
+            val node = data.get(keyRef)
+            if (node != null) {
+                cleanupEntry(node)
+            }
+        }
+    }
+
+    private fun cleanupEntry(node: KeyRefNode<K, V>) {
+        val keyRef = node.getKeyReference()
+        val key = node.getKey()
+        var nodeResurrected = false
+        var removed = false
+
+        data.computeIfPresent(keyRef) { _, newNode ->
+            // We have a new node associated with the key. Our node was already removed.
+            if (newNode != node) return@computeIfPresent newNode
+
+            synchronized(newNode) {
```

Comment:
You use here both `newNode` and `node`. Since they are the same object in fact, could you use only one of them, please?